File: CoreUtilities/CalcPad/CalcPadHTML/Script/CalcPadCode.js
// Copyright(c) Lester J. Clark and Contributors.
// Licensed under the MIT License.
// CalcPadCode.js
// <script src="Script/LJCCommon.js"></script>
// <script src="Script/TextAreaCode.js"></script>

// CalcPad functions.
class CalcPadCode
{
  // The defined items.
  #Items = [];

  // Initializes an object instance.
  constructor()
  {
  }

  // Public Methods
  // ---------------

  // Perform the calculations.
  DoCalcs(text)
  {
    let success = true;
    if (null == text)
    {
      success = false;
    }

    let lines = [];
    if (success)
    {
      lines = text.split("\n");
      if (null == lines
        || !Array.isArray(lines)
        || lines.length < 1)
      {
        success = false;
      }
    }

    if (success)
    {
      for (let index = 0; index < lines.length; index++)
      {
        let line = lines[index].trim();

        // If blank line or comment.
        if (!LJC.HasValue(line)
          || line.trim().startsWith("//"))
        {
          continue;
        }

        let refLineType = new RefLineType("None");
        let lineItemRef = this.#ParseItem(line, refLineType);
        switch (refLineType.getName())
        {
          case "assignment":
            this.#DoCalc(lineItemRef);
            break;

          case "value":
            text = this.#ClearValue(line);
            let itemRef = this.#GetItem(lineItemRef.ValueName);
            let value = this.#DoRound(lineItemRef.Rounding, itemRef.Value);
            text += ` (${value})`;

            lines[index] = text;
            let calcCode = new TextAreaCode(calcPad);
            calcCode.ReplaceLine(index, text);
            break;
        }
      }
    }
  }

  // Line Parse Methods
  // ---------------

  // Parses a line into a line item.
  #ParseItem(line, refLineType)
  {
    let retValue = null;  // Item

    refLineType.setName("None");
    let tokens = this.#SplitChars(" ", line);
    if (tokens.length > 0)
    {
      retValue = this.#ParseName(tokens, refLineType);
      if (null == retValue)
      {
        retValue = this.#ParseFormula(tokens, refLineType);
      }

      if (null == retValue)
      {
        retValue = this.#ParseAssignment(tokens, refLineType);
      }

      if (null == retValue)
      {
        retValue = this.#ParseValue(tokens, refLineType);
      }
    }
    return retValue;
  }

  // Creates a Assignment line item.
  #ParseAssignment(tokens, refLineType)
  {
    let retValue = null; // Item

    refLineType.setName("None");
    if (tokens != null
      && tokens.length > 1
      && ("=" == tokens[1]))
    {
      refLineType.setName("Assignment");
      let tokenCount = tokens.length;
      let line = tokens.join(" ");
      let refSuccess = new RefBool(true);

      // Name = name/value operator name/value.
      if (tokenCount < 3
        || tokenCount > 5)
      {
        success.Value = false;
        Error.TokenCount(line, Error.SyntaxAssignment());
      }

      // Check assignment target name.
      if (refSuccess.Value)
      {
        let token = tokens[0];
        retValue = new Item(token, 0);
        if (!"string" == typeof token[0])
        {
          success.Value = false;
          Error.NameFormat(line, token);
        }
      }

      // Get first value.
      let firstValue = 0.0;
      if (refSuccess.Value
        && tokenCount > 2)
      {
        let token = tokens[2];
        firstValue = this.#GetItemValue(line, token, refSuccess
          , Error.SyntaxAssignment());
        if (3 == tokenCount)
        {
          retValue.Value = firstValue;
        }
      }

      // Get operation.
      let operation = null;
      if (refSuccess.Value
        && tokenCount > 3)
      {
        let token = tokens[3];
        operation = this.#GetOperation(line, token, refSuccess);
      }

      if (refSuccess.Value
        && tokenCount > 4)
      {
        // Get second value.
        let token = tokens[4];
        let secondValue = this.#GetItemValue(line, token
          , refSuccess, Error.SyntaxAssignment());
        retValue.Value = this.#Calc(firstValue, operation, secondValue);
      }

      if (refSuccess.Value)
      {
        retValue = this.#SaveItem(retValue, refLineType);
      }
    }
    return retValue;
  }

  // Creates a Formula line item.
  #ParseFormula(tokens, refLineType)
  {
    let retValue = null;  // Item

    refLineType.setName("None");
    if (tokens != null
      && tokens.length > 1
      && "is" == tokens[1].toLowerCase())
    {
      refLineType.setName("Formula");
      let tokenCount = tokens.length;
      let line = tokens.join(" ");
      let success = true;

      // Name is # Name
      if (tokenCount < 4
        || tokenCount > 4)
      {
        success = false;
        Error.TokenCount(line, Error.SyntaxFormula());
      }

      retValue = new Item(tokens[0], 0);
      let value = 0;
      if (tokenCount > 2)
      {
        let token = tokens[2];
        let pattern = /[0-9\-.]+/;
        if (pattern.test(token))
        {
          value = Number(token);
        }
        else
        {
          if (!"string" == typeof token[0])
          {
            success = false;
            Error.NameFormat(line, token);
          }
        }

        if (0 == value)
        {
          let findItem = this.#GetItem(token);
          if (null == findItem)
          {
            success = false;
            Error.FormulaValue(line);
          }
          else
          {
            value = findItem.Value;
          }
        }
        retValue.FormulaValue = value;
      }

      retValue.FormulaName = "";
      if (tokenCount > 3)
      {
        retValue.FormulaName = tokens[3];
      }

      if (success)
      {
        retValue = this.#SaveItem(retValue, refLineType);
      }
    }
    return retValue;
  }

  // Creates a Name line item.
  #ParseName(tokens, refLineType)
  {
    let retValue = null;  // Item

    refLineType.setName("None");
    if (tokens != null
      && tokens.length > 0)
    {
      for (let index = 0; index < tokens.length; index++)
      {
        let token = tokens[index];

        // Name/AltName
        if (token.includes("|"))
        {
          refLineType.setName("Name");
          retValue = new Item(null, 0);

          // Remove extra spaces.
          let refSuccess = new RefBool(true);
          let text = tokens.join("");
          let names = this.#SplitChars("|", text);
          if (0 == names.length)
          {
            refSuccess.Value = false;
            let line = tokens.join(" ");
            Error.Declaration(line);
          }

          if (refSuccess.Value)
          {
            retValue.Name = names[0];
            if (names.length > 1
              && LJC.HasValue(names[1]))
            {
              retValue.AltName = names[1];
            }
            retValue = this.#SaveItem(retValue, refLineType);
          }
          index = tokens.length;
        }
      }
    }
    return retValue;
  }

  // Creates a Display Value line item.
  #ParseValue(tokens, refLineType)
  {
    let retValue = null;  // Item

    refLineType.setName("None");
    if (tokens != null
      && tokens.length > 1
      && "value:" == tokens[0].toLowerCase())
    {
      refLineType.setName("Value");
      let tokenCount = tokens.length;
      let line = tokens.join(" ");

      // Value: Name
      if (tokenCount < 2
        || tokenCount > 4)
      {
        Error.TokenCount(line, Error.SyntaxValue());
      }

      retValue = new Item(null, 0);
      if (tokenCount > 1)
      {
        let token = tokens[1];
        retValue.ValueName = token;
        let findItem = this.#GetItem(token);
        if (null == findItem)
        {
          Error.ItemNotFound(token, line);
        }
      }

      if (tokenCount > 2)
      {
        retValue.Rounding = tokens[2];
      }
    }
    return retValue;
  }

  // Other Parse Methods
  // ---------------

  // Clears the old value.
  #ClearValue(line)
  {
    let retValue = "";

    if ("string" == typeof line)
    {
      retValue = line;
      let valueIndex = retValue.indexOf("(");
      if (valueIndex > 0)
      {
        retValue = retValue.substring(0, valueIndex - 1);
      }
    }
    return retValue;
  }

  // Gets the token or data item value.
  #GetItemValue(line, token, refSuccess, syntax = null)
  {
    let retValue = -1.0;

    let refRetNumber = new RefNumber(-1.0);
    if (refSuccess.Value = this.#IsItemOrNumber(line, token, refRetNumber))
    {
      retValue = refRetNumber.Value;
      if (retValue < 0)
      {
        let findItem = this.#GetItem(token);
        if (null == findItem)
        {
          success.Value = false;
          Error.ItemNotFound(line, token, syntax);
        }
        else
        {
          retValue = findItem.Value;
        }
      }
    }
    return retValue;
  }

  // Checs if token is an operation.
  #GetOperation(line, token, refSuccess)
  {
    let retValue = "";

    refSuccess.Value = true;
    if ("+-*/".includes(token))
    {
      retValue = token;
    }
    else
    {
      refSuccess.Value = false;
      Error.Operation(line);
    }
    return retValue;
  }

  // Check if token is item name or number.
  //#IsItemOrNumber(line, token, refNumber, syntax = null)
  #IsItemOrNumber(line, token, refNumber)
  {
    let retValue = true;

    if (LJC.IsNumber(token))
    {
      refNumber.Value = Number(token);
    }
    else
    {
      refNumber.Value = -1.0;

      // Does not start with letter.
      if (typeof token[0] != "string")
      {
        retValue = false;
        Error.NameFormat(line, token);
      }
    }
    return retValue;
  }

  // Other Private Methods
  // ---------------

  // Performs an assignment calculation.
  #Calc(firstValue, operation, secondValue)
  {
    let retValue = 0.0;

    if (firstValue > 0
      && secondValue > 0)
    {
      switch (operation)
      {
        case "+":
          retValue = firstValue + secondValue;
          break;
        case "-":
          retValue = firstValue - secondValue;
          break;
        case "*":
          retValue = firstValue * secondValue;
          break;
        case "/":
          retValue = firstValue / secondValue;
          break;
      }
    }
    return retValue;
  }

  // Performs the related conversion calculations.
  #DoCalc(lineItemRef)
  {
    // "LeftName" is # "RightName"

    // Calculate smaller formula totals.
    let prevItemRef = lineItemRef;
    let nextItemRef = this.#GetItem(prevItemRef.FormulaName);
    while (nextItemRef != null)
    {
      nextItemRef.Value = prevItemRef.Value * prevItemRef.FormulaValue;
      prevItemRef = nextItemRef;
      nextItemRef = this.#GetItem(nextItemRef.FormulaName);
    }

    // Calculate larger formula totals.
    // *** Begin ***
    let items = this.#GetItemsWithFormula(lineItemRef);
    for (let index = 0; index < items.length; index++)
    {
      // *** End ***
      prevItemRef = lineItemRef;
      //nextItemRef = this.#GetItemWithFormula(lineItemRef);
      nextItemRef = items[index];
      while (nextItemRef != null)
      {
        nextItemRef.Value = prevItemRef.Value / nextItemRef.FormulaValue;
        prevItemRef = nextItemRef;
        nextItemRef = this.#GetItemWithFormula(nextItemRef);
      }
    }
  }

  // Performs the rounding.
  #DoRound(rounding, value)
  {
    let retValue = value;

    if (LJC.HasValue(rounding))
    {
      switch (rounding.toLowerCase())
      {
        case "ceiling":
          retValue = Math.ceil(value);
          break;

        case "round":
          retValue = Math.round(value);
          break;

        case "floor":
          retValue = Math.floor(value);
          break;
      }

      if (rounding.toLowerCase().startsWith("round:"))
      {
        let tokens = rounding.split(":");
        let text = value.toFixed(tokens[1]);
        retValue = parseFloat(text);
      }
    }
    return retValue;
  }

  // Gets the Item by Name or AltName
  #GetItem(name)
  {
    let retValue = null;  // Item;

    if (LJC.HasValue(name))
    {
      name = name.toLowerCase();
      retValue = this.#Items.find(x => x.Name.toLowerCase() == name);
      if (null == retValue)
      {
        retValue = this.#Items.find(x => x.AltName.toLowerCase() == name);
      }
    }
    return retValue;
  }

  // Gets all where the Formula name = item's Name or AltName.
  #GetItemsWithFormula(findItem)
  {
    let retValue = [];
    for (let index = 0; index < this.#Items.length; index++)
    {
      let success = false;
      let item = this.#Items[index];

      if (LJC.HasValue(findItem.Name))
      {
        let name = findItem.Name.toLowerCase();
        if (item.FormulaName.toLowerCase() == name)
        {
          success = true;
          retValue.push(item);
        }
      }

      if (!success
        && LJC.HasValue(findItem.AltName))
      {
        let altName = findItem.AltName.toLowerCase();
        if (item.FormulaName.toLowerCase() == altName)
        {
          retValue.push(item);
        }
      }
    }
    return retValue;
  }

  // Gets first where the Formula name = item's Name or AltName.
  #GetItemWithFormula(item)
  {
    let retValue = null;  // Item

    if (item != null)
    {
      if (LJC.HasValue(item.Name))
      {
        let name = item.Name.toLowerCase();
        retValue = this.#Items.find(x => x.FormulaName.toLowerCase() == name);
      }

      if (null == retValue
        && LJC.HasValue(item.AltName))
      {
        let altName = item.AltName.toLowerCase();
        retValue = this.#Items.find(x => x.FormulaName.toLowerCase() == altName);
      }
    }
    return retValue;
  }

  // Saves or Updates the Item.
  #SaveItem(lineItem, refLineType)
  {
    var retValue = this.#GetItem(lineItem.Name);
    if (null == retValue)
    {
      retValue = lineItem;
      this.#Items.push(retValue);
    }
    else
    {
      // Update Item values.
      switch (refLineType.getName())
      {
        case "assignment":
          retValue.Value = lineItem.Value;
          break;

        case "formula":
          retValue.FormulaValue = lineItem.FormulaValue;
          retValue.FormulaName = lineItem.FormulaName;
          break;

        case "name":
          retValue.AltName = lineItem.AltName;
          break;

        case "value":
          retValue.Rounding = lineItem.Rounding;
          break;
      }
    }

    if (!LJC.HasValue(retValue.FormulaName))
    {
      retValue.FormulaName = "";
    }
    return retValue;
  }

  // Performs a split with no empty entries.
  #SplitChars(separators, text)
  {
    if (separators == " ")
    {
      separators = "\\s+";
    }
    else
    {
      separators = `[${separators}]`;
    }
    let regex = new RegExp(separators, "g");
    let retValue = text.split(regex);
    return retValue;
  }
}

class RefLineType
{
  #Name = "";

  // Initializes an object instance.
  constructor(name = "None")
  {
    this.setName(name);
  }

  getName()
  {
    return this.#Name;
  }

  setName(name)
  {
    let value = name.toLowerCase();
    if (value == "assignment"
      || value == "formula"
      || value == "name"
      || value == "none"
      || value == "value")
    {
      this.#Name = value;
    }
  }
}

// Represents a Number that can be modified in a called method.
class RefNumber
{
  // Initializes an object instance.
  constructor(value = 0.0)
  {
    if (LJC.IsNumber(value))
    {
      this.Value = value;
    }
  }
}

// Represents a boolean value that can be modified in a called method.
class RefBool
{
  // Initializes an object instance.
  constructor(value = true)
  {
    if ("boolean" == typeof value)
    {
      this.Value = value;
    }
  }
}

// Represents a Data Item.
class Item
{
  // Initializes an object instance.
  constructor(name = "", value = 0.0)
  {
    // The Item name.
    if (LJC.HasValue(name))
    {
      this.Name = name;
    }

    // The Item value.
    if (LJC.IsNumber(value))
    {
      this.Value = value;
    }

    // The Altername name.
    this.AltName = "";

    // The Formula value.
    this.FormulaValue = 0.0;

    // The Formula Item name.
    this.FormulaName = "";

    // The Rounding value.
    this.Rounding = "";

    // The Value Item name.
    this.ValueName = "";
  }

  // Methods
  // ---------------

  // Gets int value.
  ValueIfInt(refNumber)
  {
    let retValue = false;

    refNumber.Value = 0;
    if (0 == Value % 1)
    {
      refNumber.Value = this.Value; // (int)
      retValue = true;
    }
    return retValue;
  }
}

File: CoreUtilities/CalcPad/CalcPad/CalcPadCode.cs
using LJCNetCommon;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using static System.Net.Mime.MediaTypeNames;

namespace CalcPad
{
  public class CalcPadCode
  {
    // Initializes an object instance.
    public CalcPadCode()
    {
      Items = new List<Item>();
    }

    #region Public Methods
    public void DoCalcs(RichTextBox rtb)
    {
      RTB = rtb;
      var lines = rtb.Lines;
      for (int index = 0; index < lines.Count(); index++)
      {
        var line = lines[index].Trim();

        // If blank line or comment.
        if (!NetString.HasValue(line)
          || line.Trim().StartsWith("//"))
        {
          continue;
        }

        var lineItemRef = ParseItem(line, out LineType lineType);
        switch (lineType)
        {
          case LineType.Assignment:
            DoCalc(lineItemRef);
            break;

          case LineType.Value:
            var text = ClearValue(line);
            var itemRef = GetItem(lineItemRef.ValueName);
            var value = DoRound(lineItemRef.Rounding, itemRef.Value);
            text += $" ({value})";

            ReplaceLine(RTB, index, text);
            break;
        }
      }
    }
    #endregion

    #region Line Parse Methods

    // Parses a line into a line item.
    public Item ParseItem(string line, out LineType lineType)
    {
      Item retValue = null;

      lineType = LineType.None;
      var tokens = SplitChars(" ", line);
      if (tokens.Length > 0)
      {
        retValue = ParseName(tokens, out lineType);
        if (null == retValue)
        {
          retValue = ParseFormula(tokens, out lineType);
        }

        if (null == retValue)
        {
          retValue = ParseAssignment(tokens, out lineType);
        }

        if (null == retValue)
        {
          retValue = ParseValue(tokens, out lineType);
        }
      }
      return retValue;
    }

    // Creates a Assignment line item.
    private Item ParseAssignment(string[] tokens
      , out LineType lineType)
    {
      Item retValue = null;

      lineType = LineType.None;
      if (tokens != null
        && tokens.Length > 1
        && ("=" == tokens[1]))
      {
        lineType = LineType.Assignment;
        var tokenCount = tokens.Length;
        var line = string.Join(" ", tokens);
        var success = true;

        // Name = name/value operator name/value.
        if (tokenCount < 3
          || tokenCount > 5)
        {
          success = false;
          Error.TokenCount(line, Error.SyntaxAssignment());
        }

        // Check assignment target name.
        if (success)
        {
          var token = tokens[0];
          retValue = new Item(token, 0);
          if (!char.IsLetter(token[0]))
          {
            success = false;
            Error.NameFormat(line, token);
          }
        }

        // Get first value.
        var firstValue = 0.0;
        if (success
          && tokenCount > 2)
        {
          var token = tokens[2];
          firstValue = GetItemValue(line, token, out success
            , Error.SyntaxAssignment());
          if (3 == tokenCount)
          {
            retValue.Value = firstValue;
          }
        }

        // Get operation.
        string operation = null;
        if (success
          && tokenCount > 3)
        {
          var token = tokens[3];
          operation = GetOperation(line, token, out success);
        }

        if (success
           && tokenCount > 4)
        {
          // Get second value.
          var token = tokens[4];
          var secondValue = GetItemValue(line, token
            , out success, Error.SyntaxAssignment());
          retValue.Value = Calc(firstValue, operation, secondValue);
        }

        if (success)
        {
          retValue = SaveItem(retValue, lineType);
        }
      }
      return retValue;
    }

    // Creates a Formula line Item.
    private Item ParseFormula(string[] tokens, out LineType lineType)
    {
      Item retValue = null;

      lineType = LineType.None;
      if (tokens != null
        && tokens.Length > 1
        && "is" == tokens[1].ToLower())
      {
        lineType = LineType.Formula;
        var tokenCount = tokens.Length;
        var line = string.Join(" ", tokens);
        var success = true;

        // Name is # Name
        if (tokenCount < 4
          || tokenCount > 4)
        {
          success = false;
          Error.TokenCount(line, Error.SyntaxFormula());
        }

        retValue = new Item(tokens[0], 0);
        if (tokenCount > 2)
        {
          var token = tokens[2];
          if (!double.TryParse(token, out double value))
          {
            if (!char.IsLetter(token[0]))
            {
              success = false;
              Error.NameFormat(line, token);
            }
          }

          if (0 == value)
          {
            var findItem = GetItem(token);
            if (null == findItem)
            {
              success = false;
              Error.FormulaValue(line);
            }
            else
            {
              value = findItem.Value;
            }
          }
          retValue.FormulaValue = value;
        }

        if (tokenCount > 3)
        {
          retValue.FormulaName = tokens[3];
        }

        if (success)
        {
          retValue = SaveItem(retValue, lineType);
        }
      }
      return retValue;
    }

    // Creates a Name Item.
    private Item ParseName(string[] tokens, out LineType lineType)
    {
      Item retValue = null;

      lineType = LineType.None;
      if (tokens != null
        && tokens.Length > 0)
      {
        foreach (string token in tokens)
        {
          // Name/AltName
          if (token.Contains("|"))
          {
            lineType = LineType.Name;
            retValue = new Item(null, 0);

            // Remove extra spaces.
            var success = true;
            var text = String.Join("", tokens);
            var names = SplitChars("|", text);
            if (0 == names.Length)
            {
              success = false;
              var line = string.Join(" ", tokens);
              Error.Declaration(line);
            }

            if (success)
            {
              retValue.Name = names[0];
              if (names.Length > 1)
              {
                retValue.AltName = names[1];
              }
              retValue = SaveItem(retValue, lineType);
            }
            break;
          }
        }
      }
      return retValue;
    }

    // Creates a Display Value line Item.
    private Item ParseValue(string[] tokens, out LineType lineType)
    {
      Item retValue = null;

      lineType = LineType.None;
      if (tokens != null
        && tokens.Length > 1
        && "value:" == tokens[0].ToLower())
      {
        lineType = LineType.Value;
        var tokenCount = tokens.Length;
        var line = string.Join(" ", tokens);

        // Value: Name
        if (tokenCount < 2
          || tokenCount > 4)
        {
          Error.TokenCount(line, Error.SyntaxValue());
        }

        retValue = new Item();
        if (tokenCount > 1)
        {
          var token = tokens[1];
          retValue.ValueName = token;
          var findItem = GetItem(token);
          if (null == findItem)
          {
            Error.ItemNotFound(token, line);
          }
        }

        if (tokenCount > 2)
        {
          retValue.Rounding = tokens[2];
        }
      }
      return retValue;
    }
    #endregion

    #region Other Parse Methods

    // Clears the old value.
    private string ClearValue(string line)
    {
      var retValue = line;
      var valueIndex = retValue.IndexOf("(");
      if (valueIndex > 0)
      {
        retValue = retValue.Substring(0, valueIndex - 1);
      }
      return retValue;
    }

    // Gets the token or item value.
    private double GetItemValue(string line, string token, out bool success
      , string syntax = null)
    {
      if (success = IsItemOrNumber(line, token, out double retValue))
      {
        if (retValue < 0)
        {
          Item findItem = GetItem(token);
          if (null == findItem)
          {
            success = false;
            Error.ItemNotFound(line, token, syntax);
          }
          else
          {
            retValue = findItem.Value;
          }
        }
      }
      return retValue;
    }

    // Checs if tokenis an operation.
    private string GetOperation(string line, string token, out bool success)
    {
      string retValue = null;

      success = true;
      if ("+-*/".Contains(token))
      {
        retValue = token;
      }
      else
      {
        success = false;
        Error.Operation(line);
      }
      return retValue;
    }

    // Check if token is item name or number.
    //private bool IsItemOrNumber(string line, string token
    //  , out double value, string syntax = null)
    private bool IsItemOrNumber(string line, string token
      , out double value)
    {
      var retValue = true;

      if (!double.TryParse(token, out value))
      {
        value = -1.0;

        // Not a number and does not start with letter.
        if (!Error.IsNumber(token)
          && !char.IsLetter(token[0]))
        {
          retValue = false;
          Error.NameFormat(line, token);
        }

        //if (retValue
        //  && char.IsNumber(token[0])
        //  && !Error.IsNumber(token))
        //{
        //  retValue = false;
        //  Error.Number(line, token, syntax);
        //}
      }
      return retValue;
    }
    #endregion

    #region Other Private Methods

    // Performs an assignment calculation.
    private double Calc(double firstValue, string operation
      , double secondValue)
    {
      double retValue = 0.0;

      if (firstValue > 0
        && secondValue > 0)
      {
        switch (operation)
        {
          case "+":
            retValue = firstValue + secondValue;
            break;
          case "-":
            retValue = firstValue - secondValue;
            break;
          case "*":
            retValue = firstValue * secondValue;
            break;
          case "/":
            retValue = firstValue / secondValue;
            break;
        }
      }
      return retValue;
    }

    // Performs the related conversion calculations.
    private void DoCalc(Item lineItemRef)
    {
      // "LeftName" is # "RightNames"
      // Calculate RightItem total.
      Item prevItemRef = lineItemRef;
      Item nextItemRef = null;
      if (NetString.HasValue(lineItemRef.FormulaName))
      {
        var formulaItemRef = GetItem(lineItemRef.FormulaName);
        if (formulaItemRef != null)
        {
          formulaItemRef.Value = lineItemRef.Value * lineItemRef.FormulaValue;
          nextItemRef = GetItemWithFormula(lineItemRef);
        }
      }

      if (null == nextItemRef)
      {
        // Look for related conversion.
        nextItemRef = GetItemWithFormula(lineItemRef);
      }

      // Calculate related formula totals.
      while (nextItemRef != null)
      {
        if (nextItemRef != null)
        {
          nextItemRef.Value = prevItemRef.Value / nextItemRef.FormulaValue;
          prevItemRef = nextItemRef;
          nextItemRef = GetItemWithFormula(nextItemRef);
        }
      }
    }

    // Performs the rounding.
    private double DoRound(string rounding, double value)
    {
      double retValue = value;

      if (NetString.HasValue(rounding))
      {
        switch (rounding.ToLower())
        {
          case "ceiling":
            retValue = Math.Ceiling(value);
            break;

          case "round":
            retValue = Math.Round(value);
            break;

          case "floor":
            retValue = Math.Floor(value);
            break;
        }
      }
      return retValue;
    }

    // Gets the Item by Name or AltName
    private Item GetItem(string name)
    {
      Item retValue = null;

      if (NetString.HasValue(name))
      {
        name = name.ToLower();
        retValue = Items.Find(x => 0 == string.Compare(x.Name, name, true));
        if (null == retValue)
        {
          retValue = Items.Find(x => 0 == string.Compare(x.AltName, name, true));
        }
      }
      return retValue;
    }

    // Gets the Formula name by Items Name or AltNames
    private Item GetItemWithFormula(Item item)
    {
      Item retValue = null;

      if (item != null)
      {
        if (NetString.HasValue(item.Name))
        {
          var name = item.Name.ToLower();
          retValue = Items.Find(x => x.FormulaName == name);
        }
        if (null == retValue
          && NetString.HasValue(item.AltName))
        {
          var altName = item.AltName.ToLower();
          retValue = Items.Find(x => x.FormulaName == altName);
        }
      }
      return retValue;
    }

    // Saves or Updates the Item.
    private Item SaveItem(Item item, LineType lineType)
    {
      var retValue = GetItem(item.Name);
      if (null == retValue)
      {
        retValue = item;
        Items.Add(retValue);
      }
      else
      {
        // Update Item values.
        switch (lineType)
        {
          case LineType.Assignment:
            retValue.Value = item.Value;
            break;

          case LineType.Formula:
            retValue.FormulaValue = item.FormulaValue;
            retValue.FormulaName = item.FormulaName;
            break;

          case LineType.Name:
            retValue.AltName = item.AltName;
            break;

          case LineType.Value:
            retValue.Rounding = item.Rounding;
            break;
        }
      }
      return retValue;
    }

    // Performs a split with no empty entries.
    private string[] SplitChars(string separators, string text)
    {
      return text.Split(separators.ToCharArray()
        , StringSplitOptions.RemoveEmptyEntries);
    }
    #endregion

    #region RTB Methods

    // Gets the RTB line first char index.
    private static int GetCharIndex(RichTextBox rtb, int lineIndex)
    {
      var retValue = rtb.GetFirstCharIndexFromLine(lineIndex);
      return retValue;
    }

    // Replaces RTB line.
    private static void ReplaceLine(RichTextBox rtb, int lineIndex, string text)
    {
      var beginIndex = GetCharIndex(rtb, lineIndex);

      var endIndex = rtb.Text.Length;
      if (lineIndex < rtb.Lines.Count() - 1)
      {
        lineIndex++;
        endIndex = GetCharIndex(rtb, lineIndex) - 1;
      }

      var length = endIndex - beginIndex;
      rtb.Select(beginIndex, length);
      rtb.SelectedText = text;
    }
    #endregion

    #region Properties

    // The defined Items.
    private List<Item> Items { get; set; }

    private RichTextBox RTB { get; set; }
    #endregion
  }

  // Represents a Data Item.
  public class Item
  {
    #region Constructors

    // Initializes an object instance.
    public Item()
    {
    }

    // Initializes an object instance.
    public Item(string name, double value)
    {
      Name = name;
      Value = value;
    }
    #endregion

    #region Data Methods

    // The debug display value.
    public override string ToString()
    {
      return $"{Name}: {Value}";
    }
    #endregion

    #region Methods

    // Gets int value.
    public bool ValueIfInt(out int value)
    {
      bool retValue = false;

      value = 0;
      if (0 == Value % 1)
      {
        Value = (int)Value;
        retValue = true;
      }
      return retValue;
    }
    #endregion

    #region Properties

    // The Altername name.
    public string AltName { get; set; }

    // The Formula value.
    public double FormulaValue { get; set; }

    // The Formula Item name.
    public string FormulaName { get; set; }

    // The Item name.
    public string Name { get; set; }

    // The Rounding value.
    public string Rounding { get; set; }

    // The Value Item name.
    public string ValueName { get; set; }

    // The Item value.
    public double Value { get; set; }
    #endregion
  }

  // The line type values.
  public enum LineType
  {
    Assignment,
    Formula,
    Name,
    None,
    Value
  }
}
